src/lib.rs

```diff
@@ -1,90 +1,117 @@
+pub mod error;
+pub mod support;
+pub mod consts;
+
+#[macro_use]
+mod macros;
+
 extern crate xenctrl_sys;
-use std::io::Error;
-use std::ptr::{null_mut};
-use std::os::raw::{c_void, c_ulong};
+use std::{
+    mem,
+    ptr::{null_mut, NonNull},
+};
+
+use xenctrl_sys::{
+    xc_clear_last_error, xc_domain_maximum_gpfn, xc_domain_pause, xc_domain_unpause, xc_error_code,
+    xc_get_last_error, xc_interface, xc_interface_close, xc_interface_open, xc_monitor_disable,
+    xc_monitor_enable, xentoollog_logger,
+};
 
-pub const PAGE_SHIFT: u32 = xenctrl_sys::XC_PAGE_SHIFT;
-pub const PAGE_SIZE: u32 = xenctrl_sys::XC_PAGE_SIZE;
+use error::Error;
+use support::PageInfo;
+
+type Result<T> = std::result::Result<T, Error>;
 
 #[derive(Debug)]
-pub struct Xc {
-    handle: *mut xenctrl_sys::xc_interface,
-    evtchn_port: *mut u32,
+pub struct XenControl {
+    handle: NonNull<xc_interface>,
+    evtchn_port: u32,
 }
 
-impl Xc {
-
-    pub fn new() -> Result<Self,Error> {
+impl XenControl {
+    pub fn from(
```

Comment:
Why did you replaced the `new` function, which is a convention to have an object constructor, by the name `from` ?
You break the API here, and i don't understand what this brings us ?